File: service-golang/lambda-v1-delete-task/incoming/handler.go
package incoming

import (
	"context"
	"errors"
	"fmt"
	"os"
	"root/service-golang/lambda-shared"
	"root/service-golang/lambda-v1-delete-task/core"
	"root/service-golang/lambda-v1-delete-task/outgoing"

	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/google/uuid"
	"github.com/jmoiron/sqlx"
)

type Handler struct {
	Database    *sqlx.DB
	TaskService *core.TaskService
}

func NewHandler() (*Handler, error) {
	cfg, err := config.LoadDefaultConfig(context.Background())
	if err != nil {
		return nil, fmt.Errorf("failed to load aws default config: %w", err)
	}

	secret, err := shared.GetSecret(cfg, os.Getenv("DB_SECRET_ID"))
	if err != nil {
		return nil, fmt.Errorf("failed to get database secret: %w", err)
	}

	dbConfig, err := shared.NewDatabaseConfig(secret)
	if err != nil {
		return nil, fmt.Errorf("failed to create database config: %w", err)
	}

	database, err := shared.NewDatabase(dbConfig)
	if err != nil {
		return nil, fmt.Errorf("failed to create database: %w", err)
	}

	taskRepository := outgoing.NewTaskRepository(database)
	taskService := core.NewTaskService(taskRepository)

	return &Handler{
		Database:    database,
		TaskService: taskService,
	}, nil
}

func (handler *Handler) Invoke(ctx context.Context, r events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
	taskIdString, ok := r.PathParameters["task_id"]
	if !ok {
		return events.APIGatewayProxyResponse{}, fmt.Errorf("task id not found in path parameters")
	}

	taskId, err := uuid.Parse(taskIdString)
	if err != nil {
		return events.APIGatewayProxyResponse{}, fmt.Errorf("failed to parse task id: %w", err)
	}

	err = handler.TaskService.DeleteTask(ctx, taskId)
	if errors.Is(err, outgoing.ErrTaskNotFound) {
		return events.APIGatewayProxyResponse{
			StatusCode: 404,
		}, nil
	} else if err != nil {
		return events.APIGatewayProxyResponse{}, fmt.Errorf("failed to delete task: %w", err)
	}

	return events.APIGatewayProxyResponse{
		StatusCode: 204,
	}, nil
}

File: service-java/lambda-v1-delete-task/src/main/java/science/monke/incoming/Handler.java
package science.monke.incoming;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;
import org.jdbi.v3.core.Jdbi;
import science.monke.core.TaskService;
import science.monke.outgoing.TaskRepository;

import java.util.UUID;

public class Handler
    implements RequestHandler<APIGatewayProxyRequestEvent, APIGatewayProxyResponseEvent> {

  private final TaskService taskService;

  public Handler() {
    Secret secret = Utils.getSecret(System.getenv("DB_SECRET_ID"));
    DatabaseConfig databaseConfig = new DatabaseConfig(secret);
    Jdbi jdbi = Utils.createDatabaseConnection(databaseConfig);
    TaskRepository taskRepository = new TaskRepository(jdbi);
    taskService = new TaskService(taskRepository);
  }

  @Override
  public APIGatewayProxyResponseEvent handleRequest(
      APIGatewayProxyRequestEvent apiGatewayProxyRequestEvent, Context context) {
    String taskIdString = apiGatewayProxyRequestEvent.getPathParameters().get("task_id");
    UUID taskId = UUID.fromString(taskIdString);
    int numberOfRowsDeleted = taskService.deleteTask(taskId);
    if (numberOfRowsDeleted == 0) {
      return new APIGatewayProxyResponseEvent().withStatusCode(404);
    }
    return new APIGatewayProxyResponseEvent().withStatusCode(204);
  }
}
